source/navier_stokes_projection/assemble_diffusion_rhs.cc

```diff
@@ -1,13 +1,360 @@
+              beta[0] * old_gradients[q] * old_values[q]
+              +
+              beta[1] * old_old_gradients[q] * old_old_values[q];
+      break;
+    }
+    case RunTimeParameters::ConvectiveTermWeakForm::skewsymmetric:
+    {
+      // Loop over quadrature points
+      for (std::size_t q=0; q<advection_term.size(); ++q)
+      {
+        const double old_divergence = trace(old_gradients[q]);
+        const double old_old_divergence = trace(old_old_gradients[q]);
+
+        advection_term[q] =
+              beta[0] *
+              (old_gradients[q] * old_values[q]
+               +
+               0.5 * old_divergence * old_values[q])
+              +
+              beta[1] *
+              (old_old_gradients[q] * old_old_values[q]
+               +
+               0.5 * old_old_divergence * old_old_values[q]);
+      }
+      break;
+    }
+    case RunTimeParameters::ConvectiveTermWeakForm::divergence:
+    {
+      // Loop over quadrature points
+      for (std::size_t q=0; q<advection_term.size(); ++q)
+      {
+        const double old_divergence = trace(old_gradients[q]);
+        const double old_old_divergence = trace(old_old_gradients[q]);
+
+        advection_term[q] =
+              beta[0] *
+              (old_gradients[q] * old_values[q]
+               +
+               old_divergence * old_values[q])
+              +
+              beta[1] *
+              (old_old_gradients[q] * old_old_values[q]
+               +
+               old_old_divergence * old_old_values[q]);
+      }
+      break;
+    }
+    case RunTimeParameters::ConvectiveTermWeakForm::rotational:
+    {
+      typename FEValuesViews::Vector<dim>::curl_type old_curl, old_old_curl;
+
+      // The minus sign in the argument of cross_product_2d
+      // method is due to how the method is defined.
+      if constexpr(dim == 2)
+      {
+         // Loop over quadrature points
+        for (std::size_t q=0; q<advection_term.size(); ++q)
+        {
+          old_curl[0] = old_gradients[q][1][0] - old_gradients[q][0][1];
+          old_old_curl[0] = old_old_gradients[q][1][0] - old_old_gradients[q][0][1];
+
+          advection_term[q] =
+              beta[0] * old_curl * cross_product_2d(-old_values[q])
+              +
+              beta[1] * old_old_curl * cross_product_2d(-old_old_values[q]);
+        }
+      }
+      else if constexpr(dim == 3)
+      {
+
+        // Loop over quadrature points
+        for (std::size_t q=0; q<advection_term.size(); ++q)
+        {
+          old_curl[0] = old_gradients[q][3][2] - old_gradients[q][2][3];
+          old_curl[1] = old_gradients[q][0][3] - old_gradients[q][3][0];
+          old_curl[2] = old_gradients[q][1][0] - old_gradients[q][0][1];
+          old_old_curl[0] = old_old_gradients[q][3][2] - old_old_gradients[q][2][3];
+          old_old_curl[1] = old_old_gradients[q][0][3] - old_old_gradients[q][3][0];
+          old_old_curl[2] = old_old_gradients[q][1][0] - old_old_gradients[q][0][1];
```

Comment:
If I am not mistaken, this can also be done with a `get_curl` method of the `FEValues` instance.